kernel/src/mem.rs

```diff
@@ -1,202 +1,292 @@
+    /// reclaimed, this method must return 0.
+    ///
+    /// # Default AppSlice
+    ///
+    /// A default instance of an AppSlice must return 0.
+    fn len(&self) -> usize;
 
-/// Type for specifying an AppSlice is shared with the kernel.
-#[derive(Debug)]
-pub struct Shared;
+    /// Pointer to the first byte of the userspace memory region.
+    ///
+    /// If the length of the initially shared memory region
+    /// (irrespective of the return value of [`len`](Read::len)) is 0,
+    /// this function returns a pointer to address `0x0`. This is
+    /// because processes may allow buffers with length 0 to share no
+    /// no memory with the kernel. Because these buffers have zero
+    /// length, they may have any pointer value. However, these
+    /// _dummy addresses_ should not be leaked, so this method returns
+    /// 0 for zero-length slices.
+    ///
+    /// # Default AppSlice
+    ///
+    /// A default instance of an AppSlice must return a pointer to
+    /// address `0x0`.
+    fn ptr(&self) -> *const u8;
 
-/// Base type for an AppSlice that holds the raw pointer to the memory region
-/// the app shared with the kernel.
-pub(crate) struct AppPtr<L, T> {
-    ptr: NonNull<T>,
-    process: AppId,
-    _phantom: PhantomData<L>,
+    /// Applies a function to the (read only) slice reference pointed
+    /// to by the AppSlice.
+    ///
+    /// If the process is no longer alive and the memory has been
+    /// reclaimed, this method must return the default value.
+    ///
+    /// # Default AppSlice
+    ///
+    /// A default instance of an AppSlice must return the passed
+    /// default value without executing the closure.
+    fn map_or<F, R>(&self, default: R, fun: F) -> R
+    where
+        F: FnOnce(&[u8]) -> R;
 }
 
-impl<L, T> AppPtr<L, T> {
-    /// Safety: Trusts that `ptr` points to userspace memory in `appid`
-    unsafe fn new(ptr: NonNull<T>, appid: AppId) -> AppPtr<L, T> {
-        AppPtr {
-            ptr: ptr,
-            process: appid,
-            _phantom: PhantomData,
-        }
-    }
+/// A readable and writable region of userspace memory.
+///
+/// This trait can be used to gain read-write access to memory regions
+/// wrapped in an AppSlice type.
+///
+/// This is a supertrait of [`Read`], which features further methods.
+pub trait ReadWrite: Read {
+    /// Mutable pointer to the userspace memory region
+    ///
+    /// If the length of the initially shared memory region
+    /// (irrespective of the return value of [`len`](Read::len)) is 0,
+    /// this function must return a pointer to address `0x0`. This is
+    /// because processes allow buffers with length 0 to reclaim
+    /// shared memory with the kernel and are allowed to specify _any_
+    /// address, even if it is not contained within their address
+    /// space. These _dummy addresses_ should not be leaked to outside
+    /// code.
+    fn mut_ptr(&self) -> *mut u8;
```

Comment:
This method is not used anywhere in Tock. Should we remove it? I struggle to imagine a use case for it currently. It is not inherently unsafe, because it returns a `mut` pointer rather than a reference, but it seems to encourage unsafe behavior in that the only reason to use this instead of `Read::ptr()` is if you plan to convert the pointer into a reference. 

At the very least, this method should take `&'mut self`, much like #2465 changes `mut_map_or()` to do. But I think it is better to just remove it given that it is completely unused.

Reply:
A quick, and likely imperfect, grep through the history and I can't find when this was _ever_ used. Seems sensible to remove it given non-use, can always add it back if something can make the case for its existence.